src/common/atomic_queue.cpp

```diff
@@ -559,6 +689,355 @@ void *process_queue(void *arg) {
 
             break;
         }
+
+        case ATOMIC_SCATTER_INDEX:
+            // Atomic scatter - indexed
+            if (msgPointer->flag & ATOMIC_WRITE_COMPLETED) {
+                // last write completed. If buffer is still allocated,
+                // deallocate and  remove the entry
+                if (msgPointer->flag & ATOMIC_BUFFER_ALLOCATED) {
+                    try {
+                        allocator->deallocate(ATOMIC_REGION_ID,
+                                              msgPointer->offsetBuffer);
+                        allocator->deallocate(ATOMIC_REGION_ID,
+                                              msgPointer->offsetIndex);
+                    } catch (...) {
+                        popStatus = DEALLOCATEERROR;
+                    }
+                }
+                break;
+            }
+            if (msgPointer->flag & ATOMIC_WRITE_IN_PROGRESS) {
+                // Indexed scatter- Inprogress; buffer
+                try {
+                    localPointerB = allocator->get_local_pointer(
+                        ATOMIC_REGION_ID, msgPointer->offsetBuffer);
+                } catch (...) {
+                    popStatus = MAPERROR;
+                }
+            } else {
+                // New request; allocate buffer and read from client
+                uint64_t offsetB = 0;
+                uint64_t bufferSize =
+                    msgPointer->inElements * msgPointer->indexSize;
+                try {
+                    offsetB = allocator->allocate(ATOMIC_REGION_ID, bufferSize);
+                    // Update the message with the region and offset of buffer
+                    msgPointer->offsetBuffer = offsetB;
+                    msgPointer->flag |= ATOMIC_BUFFER_ALLOCATED;
+                    // Read from client into buffer
+                    try {
+                        localPointerB = allocator->get_local_pointer(
+                            ATOMIC_REGION_ID, offsetB);
+                        try {
+                            retStatus = fabric_read(
+                                msgPointer->key, localPointerB, bufferSize, 0,
+                                fiAddr,
+                                famOpsLibfabricQ->get_defaultCtx(uint64_t(0)));
+                            // Set the flag to indicate write is in progress
+                            msgPointer->flag |= ATOMIC_WRITE_IN_PROGRESS;
+                            try {
+                                openfam_persist(localPointerB, bufferSize);
+                                openfam_persist(msgPointer,
+                                                sizeof(msgPointer->flag));
+                            } catch (...) {
+                                retStatus = PERSISTERROR;
+                            }
+                            try {
+                                openfam_persist(localPointerB, bufferSize);
```

Comment:
You can combine try-catch for all openfam_persist calls, unless you want to perform some cleanup operation for any of these call failures.